### PR TITLE
148 Docker image (Web + Cron + PHP)

### PR DESCRIPTION
**Pull request by saschahemleb**
This PR will extend our application docker image:
- NGINX, so we can get rid of the nginx service in docker-compose.yml
- crond, now laravels scheduler should work ootb
- php-fpm (as before)
- supervisord as our init

There are a few gotchas:
- Because we now include a .dockerignore, a few host files aren't included in the image anymore. Most prominent is the .env file, which was used by the docker-compose setup to set the APP_ENV, APP_KEY etc.
It's probably fine to just address it in the README accordingly

- The image is now build in multiple stages and I pulled all the npm stuff into their own stage. Upside is that we do not need the npm tools in our final image. But if someone or something was relying on those tools they are out of luck now.

### CODE REVIEW

File: docker_boot.sh [hunk 1]
```diff
@@ -1,5 +1,14 @@
...
-./wait-for-it.sh database2:3306 -- php artisan migrate
+set -e
 
-php-fpm
+DB_HOST="${DB_HOST:?environment variable is missing}"
+DB_PORT="${DB_PORT:-3306}"
+
+./wait-for-it.sh "${DB_HOST}:${DB_PORT}" -- php artisan migrate --force
+
+# php artisan optimize
```
Comment (range-of-motion) [inline docker_boot.sh:10]: What is this supposed to do?

File: docker_boot.sh [hunk 2]
```diff
@@ -1,5 +1,14 @@
...
+
+php artisan storage:link
+
+exec $@
```
Comment (range-of-motion) [inline docker_boot.sh:14]: What is this supposed to do?

File: Dockerfile [hunk 3]
```diff
@@ -1,31 +1,62 @@
...
-# Required before we're able to run any "php artisan" commands
-RUN composer install
+COPY composer.json composer.lock ./
+RUN composer install --no-dev --no-scripts --no-autoloader
+
+COPY --chown=www-data . ./
+COPY --chown=www-data --from=assets /build/public/ ./public/
 
-RUN php artisan budget:install
+RUN composer dumpautoload -oa
```
Comment (range-of-motion) [inline Dockerfile:58]: Why is this required after `composer install`?

Comment (saschahemleb) [reply Dockerfile:58]: Because I didn't dump the autoloader during installation with ```--no-autoloader```.

Initially I only copy the composer.json and .lock files into the container and run the installation.   
This way that layer won't get rebuild if some changes in the actual source code is made; it will only rebuild if the actual composer files get changed.

File: docker_boot.sh (hunk 2 above)
Comment (saschahemleb) [reply docker_boot.sh:14]: This will replace the current process with whatever is passed to the entrypoint.
So the Dockerfile sets the entrypoint to this docker_boot.sh script and the cmd is the actual supervisord in blocking mode.

This just mimics what is done in the upstream package and their docker-entrypoint.sh. If we have no use for overriding the CMD we could hardcode the supervisord command here

File: docker_boot.sh (hunk 1 above)
Comment (saschahemleb) [reply docker_boot.sh:10]: The optimize command precompiles the config, routes, services and providers so they don't need to be parsed for every request. For example in config files there are a few calls to the env(), but these do not change after starting the container so the resolved values can be cached. Just a little optimization.

But when I tried to run it, it failed because there was a route with an anonymous function, and those can't be cached. So I commented out the call.

Comment (range-of-motion) [reply docker_boot.sh:10]: Ah, right. I recently read on Taylor's Twitter that closures will be cacheable in a future version of Laravel, but that doesn't do shit for us right now 😂

**Comment by range-of-motion** [comment]
Is this a functional Docker image? I haven't tried it yet, but it seems promising enough to try and get it set up.

Also, take a look at your commits, I think the author is fucky? 😛

**Comment by saschahemleb** [comment]
> Is this a functional Docker image? I haven't tried it yet, but it seems promising enough to try and get it set up.
> 
> Also, take a look at your commits, I think the author is fucky? 😛

Thanks for letting me know about the author, I fixed it.

And yeah, as far as I can tell everything is working with the image. I hadn't had time to test if the schedules are working, so a maybe for the crond stuff. But other than that the image is good to go!

**Comment by range-of-motion** [comment]
Trying the image right now. For some reason it keeps saying `SQLSTATE[HY000] [1045] Access denied for user 'forge'@'localhost' (using password: NO)`, while I specify the `DB_HOST` explicitly (as seen below). 

`docker build --no-cache -t testing_sascha . && docker run --env DB_HOST=host.docker.internal -it testing_sascha`

Any ideas? Can you walk me through the steps it took to get it working for you?

**Comment by saschahemleb** [comment]
You are missing the other environment variables for the database connection, for example DB_USERNAME and DB_PASSWORD. If you want to run it standalone you have to provide those too.

Another way would be to use the docker-compose.yml, that's how I tested it. Make sure the .env file is loaded by docker (which is configured correctly if you use my latest branch).

**Comment by Facinorous-420** [comment]
Is this the docker containers I should be using or the other PR?

**Comment by range-of-motion** [comment]
> Is this the docker containers I should be using or the other PR?

Depends on what you mean by "using". This is one of the candidates for the Docker image, but you *can* use it already if you want.

**Comment by range-of-motion** [comment]
Ah, thanks for your comment @saschahemleb, I got it working. I think this looks pretty good.

What am I supposed to do with the `.env`? It doesn't exist after I clone the repository, should I manually create it and run `php artisan key:generate`, or is that something the Docker image should do?

**Comment by range-of-motion** [comment]
@saschahemleb seems like this PR is the winner in the vote, and can therefor become the Docker image for this project.

* Can you modify the instructions for Docker in `README.md`?
* Do you want to keep the `docker-compose.yml`, as well as the database server that's inside?

**Comment by saschahemleb** [comment]
Hey, sounds good to me.

Regarding the `APP_KEY`/`.env`: The key should be fairly static in a live environment, so the generation during container startup is a no go, at least without persisting the key someway. For someone just cloning the repo and running a `docker-compose up` I will add a note in the README. It boils down to copying `.env.example` to `.env`, reviewing and changing any existing value (and adding a key).

We can keep the `docker-compose.yml` for now. It's a good starting point for quickly spinning up an environment.

I will push some changes to the `README` tonight as well so this can be merged soon.

File: README.md [hunk 4]
```diff
@@ -53,9 +53,18 @@ php artisan budget:update
 
 ## Docker
 
-You can run Budget using Docker. Spinning it up (`docker-compose up`) will create 3 containers–for MySQL, PHP and NGINX.
-
-*Disclaimer–currently the Docker setup is missing cronjobs and the queue worker.*
+You can run Budget using Docker Compose. Docker starts 2 containers - one for the database and the other for the application itself.
```
Comment (range-of-motion) [inline README.md:56]:
```diff
- .. using Docker Compose. Docker starts 2 containers ..
+ .. using Docker Compose. Docker Compose starts 2 containers ..
```

Trying to be as specific as possible, to prevent misunderstandings.

File: README.md [hunk 5]
```diff
@@ -53,9 +53,18 @@ php artisan budget:update
...
+
+1. Clone the repository.
+2. Copy the `.env.example` to `.env`, make sure to add an `APP_KEY` and review/replace any other value as you wish.\
+   The `APP_KEY` can be any random value but should be rotated carefully as session cookies are encrypted with this value.
```
Comment (range-of-motion) [inline README.md:60]: Is this truly the case? There are no length or format requirements for `APP_KEY`?

File: README.md [hunk 6]
```diff
@@ -53,9 +53,18 @@ php artisan budget:update
...
+3. Make sure you have the most up-to-date images.\
+`docker-compose pull`
```
Comment (range-of-motion) [inline README.md:62]: For simplification I would just put the lines next to each other, like below. But that's up to you 🙂 

```diff
- 3. Make sure you have the most up-to-date images.\
- `docker-compose pull`
+ 3. Make sure you have the most up-to-date images (`docker-compose pull`)
```

File: README.md (hunk 5 above)
Comment (saschahemleb) [reply README.md:60]: Good catch. I dug a little deeper: There is no format requirement but the length is important.\
For the cipher in our config `AES-256-CBC` this has to be 32 chars long.\
I will change the phrasing accordingly

**Comment by recklessnl** [comment]
Great work @saschahemleb. 

Would this also allow users to set up PGID and PUID as env variables, for easy permission mapping to bind mounts? The way LSIO does it is by far the best method in my experience: https://docs.linuxserver.io/general/understanding-puid-and-pgid

**Comment by saschahemleb** [comment]
> Great work @saschahemleb.
> 
> Would this also allow users to set up PGID and PUID as env variables, for easy permission mapping to bind mounts? The way LSIO does it is by far the best method in my experience: https://docs.linuxserver.io/general/understanding-puid-and-pgid

Hey @recklessnl,

maybe I'm missing something but what are you binding into the container that is used by budget?

I intentionally left any user/group id stuff out of this as I saw no use for that.

File: Dockerfile (hunk 3 above)
Comment (range-of-motion) [reply Dockerfile:58]: Gotcha, smart.

File: README.md (hunk 5 above)
Comment (range-of-motion) [reply README.md]: Cool, thanks.

**Comment by range-of-motion** [comment]
I think this is almost done. Sorry it took so long for me to review it.

It took it for another spin tonight, and it mostly works. I think the remaining to-do's are as follows.

* There is an error about the version of nginx-light not being found (do you have that too?)
* Perhaps a rebase on master, but I don't know if that's truly necessary 🤷 
* Clean up instructions for README.md (remove any dots and backslashes at the end of the lines, so it's the same as the other instructions in that file)
* Squash commits into a few (or single) commit (something like "Rework Dockerfile", whatever)

If anyone has something to say that I missed, please speak up 🙂

**Comment by saschahemleb** [comment]
> There is an error about the version of nginx-light not being found (do you have that too?)

Yeah I get that too. This specific version of the nginx-light package does not exist in the debian repository anymore. I'll remove the version constraints for now. Debian itself wouldn't allow major version upgrades in the same release so there is no need for us to force the package version.

**Comment by range-of-motion** [comment]
It looks pretty decent to me, I'm thinking of accepting it. 

Anyone else that has something to say?

File: README.md [hunk 7]
```diff
@@ -53,9 +53,17 @@ php artisan budget:update
 
 ## Docker
 
-You can run Budget using Docker. Spinning it up (`docker-compose up`) will create 3 containers–for MySQL, PHP and NGINX.
+You can run Budget using Docker Compose. Docker Compose starts 2 containers - one for the database and the other for the application itself.
 
-*Disclaimer–currently the Docker setup is missing cronjobs and the queue worker.*
+1. Clone the repository
+2. Copy the `.env.example` to `.env`, make sure to review/replace any other value as you wish
```
Comment (Facinorous-420) [inline README.md:59]: Maybe we should detail exactly what configuration options are needed for the project to run at a minimal.

**Review by Facinorous-420 — COMMENTED** [review]
Running through the README to see if it was as simple to run as possible, I came across this error: https://bss.nz/i/10-03-2020-xjbvtvJlO6.png

This is doing exactly what the README said to do, leaving everything in the .env the way it was besides the APP_URL line, changed that to my localip for the machine and the APP_KEY as instructed in step 3.
I changed nothing else.

File: docker-compose.yml [hunk 8]
```diff
@@ -19,20 +18,12 @@ services:
 
   php:
     build: ./
+    ports:
+      - 8080:80
+    env_file:
+    - .env
```
Comment (range-of-motion) [inline docker-compose.yml:24]: Can you add 2 spaces for indentation?

File: docker-compose.yml [hunk 9]
```diff
@@ -19,20 +18,12 @@ services:
 
   php:
```
Comment (range-of-motion) [inline docker-compose.yml:19]: Can we rename this to "app" or something? Since it's no longer *just* the PHP container, but also the webserver.

File: README.md [hunk 10]
```diff
@@ -53,9 +53,17 @@ php artisan budget:update
...
+3. Generate an `API_KEY` (`docker-compose run --entrypoint php php artisan key:generate --no-ansi --show`) and put that into your `.env`
+   The `APP_KEY` should be rotated carefully as session cookies are encrypted with this value
+3. Make sure you have the most up-to-date images (`docker-compose pull`)
+4. Run the stack (`docker-compose up`)
```
[the remark this answers is not in the document]
Comment (range-of-motion) [reply README.md:63]: I think this is reasonable. Or describe it as an option.

File: Dockerfile [hunk 11]
```diff
@@ -1,31 +1,59 @@
...
+    supervisor \
+    cron \
+    libzip-dev \
+    unzip \
+ && rm -rf /var/lib/apt/lists/*
+
+# Install PHP extensions
+RUN docker-php-ext-install pdo_mysql \
+ && docker-php-ext-install zip \
+ && docker-php-ext-install calendar
```
Comment (range-of-motion) [inline Dockerfile:37]: No big deal, but can make sure we use the same level of indentation here as on lines 27, 28, etc?

**Comment by range-of-motion** [comment]
Going to close this as it's a bit stale, and we've had a Docker setup for a while now. That being said, I greatly appreciate the time and effort everybody put in here. Thanks 🙏
